src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,219 @@
+ * IntersectionObserverPolyfill will
+ * create a PositionObserver class (Not implemented yet) to track the
+ * intersection between element and viewport.
+ * The IntersectionObserver receives a callback function and an optional option
+ * as params. If option is not provided, the default thresholds value will be
+ * used. Whenever the element intersection ratio cross a threshold value,
+ * IntersectionObserverPolyfill will call the callback function.
+ * Note: To share the IntersectionObserverEntry to an iframe, whoever create the
+ * IntersectionObserverPolyfill needs to handle postMessageApi and takes care of
+ * sending messages rate etc outside the IntersectionObserverPolyfill.
+ */
+export class IntersectionObserverPolyfill {
+  /**
+   * @param {!function()} callback.
+   * @param {Object=} opt_option
+   */
+  constructor(callback, opt_option) {
+    /**
+     * @private @const {function()}
+     */
+    this.callback_ = callback;
+
+    /**
+     * TODO: Need to polyfill root/rootMargin?
+     * A list of threshold, sorted in increasing numeric order
+     * @private @const {!Array}
+     */
+    this.threshold_ = opt_option ? opt_option.threshold : DEFAULT_THRESHOLD;
+
+    /** @private {?Element} */
+    this.element_ = null;
+
+    /** @private @const {number} */
+    this.prevTime_ = -1;
+
+    /**
+     * The prev threshold slot which the previous ratio fills
+     * Range [0, this.threshold_.length]
+     * TODO: Do we always want to call callback at the first tick?
+     * @private {number}
+     */
+    this.prevThresholdSlot_ = 0;
+
+    // TODO: Add the PositionObserver here.
+
+    // TODO: Add the postMessageApi logic outside IntersectionObserverPolyfill.
+  }
+
+  /**
+   * Provide a way to observer the intersection change for a specific element
+   * @param {!Element} element
+   */
+  observe(element) {
+    // TODO: create the PositionObserver class, or create the listener for
+    // position info if inside an iframe.
+    this.element_ = element;
+  }
+
+  /**
+   * Tick function that PositionObserver can tell the class to calculate
+   * intersection.
+   * @param {!./layout-rect.LayoutRectDef} element element's rect
+   * @param {!./layout-rect.LayoutRectDef} viewport viewport's rect.
+   */
+  tick(element, viewport) {
+    const changeEntry =
+        this.getValidIntersectionChangeEntry_(element, viewport);
+    if (changeEntry) {
+      this.callback_(changeEntry);
+    }
+  }
+
+  /**
+   * Return a change entry for that should be compatible with
+   * IntersectionObserverEntry if it's valid with current config.
+   *
+   * Mutates passed in rootBounds to have x and y according to spec.
+   *
+   * @param {!./layout-rect.LayoutRectDef} element element's rect
+   * @param {!./layout-rect.LayoutRectDef} viewport viewport's rect.
+   * @return {?IntersectionObserverEntry} A valid change entry.
+   * @private
+   */
+  getValidIntersectionChangeEntry_(element, viewport) {
```

Comment:
This will break flying carpets! 😢

We absolutely need to calculate offsets using the Owner element, if one exists.